Integrate new platform (core) server side into Kibana

azasypkin opened the pull request:
In this PR we're integrating base parts of the new platform (later `core`) server side into Kibana. At this point there are only two major touch points between new `core` and the rest of Kibana:

* Now it's the `core` who listens on HTTP ports, handles TLS configuration and base path proxy. All requests to Kibana will hit HTTP server exposed by the `core` first and it will decide whether request can be solely handled by the new platform or request should be forwarded to the legacy Kibana. Technically we have two Hapi servers now (one for `core` and one for legacy Kibana), but only one underlying Node HTTP listener. This setup allows us to _gradually_ introduce any "pre-route" processing code, expose new routes or replace old ones handled by the legacy platform currently.
* Once config has been validated by the legacy Kibana, part of it will also be validated by the new `core` so that we can make config validation stricter with the new config validation system (even though it's based on Joi internally we introduced custom rules tailored to our needs like `byteSize` or `duration`). That means you can notice new config validation error messages if config value has been accepted by legacy config validation code, but rejected by the `core`.

With this PR legacy Kibana will still be started first using existing CLI and will bootstrap `core` only when needed, but we're going to inverse this in https://github.com/elastic/kibana/pull/19994.  

-----------------------

New platform formerly resided in [`new-platform` feature branch](https://github.com/elastic/kibana/tree/new-platform) and then major part of it has been migrated to [`kbn-core` feature branch](https://github.com/elastic/kibana/tree/kbn-core). `new-platform` feature branch is going to stay for a while so that we can migrate the missing parts when we're ready. Here is a list of notable `kbn-core` vs `new-platform` differences:

* Removed [plugins supporting code](https://github.com/elastic/kibana/tree/new-platform/platform/src/server/plugins), [Elasticsearch client](https://github.com/elastic/kibana/tree/new-platform/platform/src/server/elasticsearch), [cli](https://github.com/elastic/kibana/tree/new-platform/platform/src/cli)
* Reorganized `platform` (now `core`) directory structure according to https://github.com/elastic/kibana/issues/12466
* Made [kbn-observable](https://github.com/elastic/kibana/tree/new-platform/packages/kbn-observable) package a part of the `core` and moved it to `src/core/lib` to use relative paths, removed unused operators
* Transformed `@kbn/utils` into `src/core/server/config/schema` and based it on Joi internally

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2892/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3165/)

azasypkin commented:
Jenkins, test this

azasypkin commented:
Jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3311/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3381/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4059/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4072/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4074/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4755/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5226/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5477/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5522/)

azasypkin commented:
retest (seems that failing security functional tests are not related to kbn-core).

azasypkin commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5534/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5828/)

archanid commented:
Question about the kbn-observable that we changed into src/core/lib/kbn_observable: why didn't we move the share_last operator over? Should we make a note about this, in case we plan to reimplement that? I assume we don't want it anymore at all?

epixa commented:
@archanid Are you referring to kbn-observable functionality from the `new-platform` branch that wasn't included here? If so, that shouldn't block this change.

The `new-platform` branch was abandoned the moment we created `kbn-core`, it continues to exist only so we can reference and potentially scavenge pieces from it in the future if we need to. So if there is a feature from it that you'd like to use going forward, you can copy it from that branch and PR to master just like any other change.

archanid commented:
@epixa yes, indeed.
Hm, but why didn't we include everything but that one operator? I just wonder. The description says we pulled the whole kbn-observable lib in here. Is it missing, or deliberately omitted?

epixa commented:
@archanid Our stance with code has always been to delete it if it's not being used, and it's not being used. I'm not sure if that's why it was omitted, but it's why I would have omitted it had I made the change :)

azasypkin commented:
> Question about the kbn-observable that we changed into src/core/lib/kbn_observable: why didn't we move the share_last operator over? Should we make a note about this, in case we plan to reimplement that? I assume we don't want it anymore at all?

I just removed everything we don't need in the initial merge, and it was the only operator that depended on RxJS. Since we have this in `new-platform` (as well as other pieces like elasticsearch client, plugins support etc.) we can pull it in when/if we need it.

epixa commented:
@archanid woo!

@azasypkin Please update the title/description and fix the dependency conflicts and get this beautiful beast merged!

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6606/)

azasypkin commented:
6.x/6.4: https://github.com/elastic/kibana/commit/f88d0b92a2aa3d36d069092c10782660799375ca

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6612/)

azasypkin commented:
Hmm, why CI is still running for this PR? :) Anyway it seems CI failed because of flaky test...

azasypkin commented:
retest

w33ble commented:
This PR broke Canvas. We just see a bunch of 404 errors in the console and nothing loads. There's an extra slash in the basepath now, not sure if that's related.

![screenshot 2018-07-13 15 00 00](https://user-images.githubusercontent.com/404731/42716984-1a320670-86b3-11e8-9513-854462173981.png)
